Add shopping cart model

JohnGrey17 opened the pull request.

JohnGrey17 committed:
initial

JohnGrey17 committed:
Created Category Entity

JohnGrey17 committed:
made relation between categories and book

JohnGrey17 committed:
Created category repository

JohnGrey17 committed:
added findAllByCategoryId to Book repository

JohnGrey17 committed:
implemented Category request/response DTO

JohnGrey17 committed:
added
BookDtoWithoutCategory
added fields to BookDto(request/response)

JohnGrey17 committed:
Created Category Mapper

JohnGrey17 committed:
moved ServiceImpl to common folder with their Interface

JohnGrey17 committed:
Created CategoryServiceImpl(added methods)
override all method (withoutImpl) to Category ServiceImpl

JohnGrey17 committed:
Created category controller with (Roles,HTTP method,Tag,)
withOut implementation

JohnGrey17 committed:
Added ChangeLog
create-category-table
add-category-id-to-books-table
add-book_id-to-categories-table
add-foreign-keys

JohnGrey17 committed:
Added ChangeLog
create-category-table
add-category-id-to-books-table
add-book_id-to-categories-table
add-foreign-keys

JohnGrey17 committed:
09-add-category-id-to-books-table.yaml
10-add-book_id-categories-table.yaml
11-add-foreing-keys.yaml
removed changelog parameters
---------------------------
added changeLog is called book_Categories

JohnGrey17 committed:
implemented solution

JohnGrey17 committed:
implemented solution

JohnGrey17 committed:
created CartItem and ShoppingCart models

JohnGrey17 committed:
created CartItemRepository and ShoppingCartRepository

JohnGrey17 committed:
commit

JohnGrey17 committed:
created ShopingCart table and CartItem in ChangeLog

JohnGrey17 committed:
created shoppingCart and cartItemDTO
created shoppingCart and cartItem mapper

JohnGrey17 committed:
changed method in ShoppingCartRepository

JohnGrey17 committed:
Updated register method in UserServiceImpl
added availability to create new Shopping cart that related to new User

JohnGrey17 committed:
Added ChangeLog 13-add-books-to books-table.yaml

JohnGrey17 committed:
commit before checkout to anotherBranch

JohnGrey17 committed:
Fixed N + 1 problem

JohnGrey17 committed:
implementing
CRUD METHODS for shopping_carts

JohnGrey17 committed:
fixed solution

HAIOVYI reviewed (approved):
GJ

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
Merge branch 'add-category-model' into Add-ShoppingCart-model

# Conflicts:
#	src/main/java/org/example/bookstore/repository/book/BookRepository.java

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
Merge branch 'master' into Add-ShoppingCart-model

# Conflicts:
#	src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java
#	src/main/java/org/example/bookstore/mapper/BookMapper.java
#	src/main/java/org/example/bookstore/model/Book.java
#	src/main/java/org/example/bookstore/service/user/UserServiceImpl.java
#	src/main/resources/db/changelog/db.changelog-master.yaml

JohnGrey17 committed:
fixed conflicts

yevheniisynytsia commented on src/main/java/org/example/bookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,70 @@
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @Operation(summary = "Delete Cart Item", description = "Delete a cart item "
+            + "with the given ID.")
+    public void deleteCartItemById(@PathVariable Long cartItemId) {
```
```suggestion
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    public void deleteCartItemById(@PathVariable Long cartItemId) {
```

yevheniisynytsia commented on src/main/java/org/example/bookstore/dto/cartitemdto/CartItemUpdateDto.java:
```diff
@@ -0,0 +1,11 @@
+    private String bookTitle;
+    @Min(1)
+    private Integer quantity;
+}
```
![image](https://github.com/JohnGrey17/bookStore/assets/77681963/58191615-5fe4-41bb-9456-a7b47ffdf0b3)

yevheniisynytsia commented on src/main/java/org/example/bookstore/mapper/CartItemMapper.java:
```diff
@@ -0,0 +1,43 @@
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+
+    @Mapping(target = "bookId", source = "book", qualifiedByName = "bookIdByBook")
```
```suggestion
    @Mapping(target = "bookId", source = "book.id")
```
this should work

yevheniisynytsia commented on src/main/java/org/example/bookstore/mapper/CartItemMapper.java:
```diff
@@ -0,0 +1,43 @@
+public interface CartItemMapper {
+
+    @Mapping(target = "bookId", source = "book", qualifiedByName = "bookIdByBook")
+    @Mapping(target = "bookTitle", source = "book", qualifiedByName = "getTitleByBook")
```
```suggestion
    @Mapping(target = "bookTitle", source = "book.title")
```

yevheniisynytsia commented on src/main/java/org/example/bookstore/model/Book.java:
```diff
@@ -41,7 +41,7 @@ public class Book {
     @Column (name = "cover_image")
     private String coverImage;
-    @ManyToMany(fetch = FetchType.LAZY)
+    @ManyToMany(fetch = FetchType.EAGER)
```
EAGER is a bad solution. Let's remember how to use LAZY and save its efficiency

yevheniisynytsia commented on src/main/java/org/example/bookstore/model/Book.java:
```diff
@@ -41,7 +41,7 @@ public class Book {
     private String description;
```
please add `@EqualsAndHashCode` to your entities

yevheniisynytsia commented on src/main/java/org/example/bookstore/model/CartItem.java:
```diff
@@ -0,0 +1,37 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public ShoppingCartResponseDto getUserShoppingCartById(Long userId, Pageable pageable) {
```
I think you don't need `Pageable` here

yevheniisynytsia commented on src/main/java/org/example/bookstore/repository/shoppingcart/ShoppingCartRepository.java:
```diff
@@ -0,0 +1,15 @@
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart,Long> {
+
+    @Query("SELECT sc FROM ShoppingCart sc LEFT JOIN FETCH sc.cartItems WHERE sc.user.id = :userId")
+    ShoppingCart findShoppingCartByUserId(@Param("userId") Long userId);
```
should be `Optional<ShoppingCart>`

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+    @Override
+    public ShoppingCartResponseDto getUserShoppingCartById(Long userId, Pageable pageable) {
+        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId);
+        return shoppingCartMapper.toDto(shoppingCartByUserId);
```
there should be a `null` check

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+        if (shoppingCart == null) {
+            throw new DataIntegrityViolationException("Shopping cart "
+                    + "not found for user id: " + userId);
+        }
```
you can move this logic into a private method and reuse in `getUserShoppingCartById`

yevheniisynytsia reviewed (changes requested):
Great job! Left some comments

JohnGrey17 committed:
*added @EqualsAndHashCode to Book entity
*added (fetch = FetchType.LAZY) -> Book book
*removed Pageable from method getUserShoppingCartById (controller level)

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -39,12 +38,9 @@ public void createNewShoppingCart(User user) {
-                    + "not found for user id: " + userId);
-        }
-
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("can`t find shopping "
+                        + "can`t by id: " + userId));
```
this code is duplicated

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -58,8 +54,11 @@ public ShoppingCartResponseDto addBookToCartItem(Long userId, CartItemRequestDto
-        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId);
+    @Transactional
+    public ShoppingCartResponseDto getUserShoppingCartById(Long userId) {
+        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("can`t find shoppingCart by id: "
+                        + userId));
```
this code is duplicated
